src/plot_api/plot_api.js

```diff
@@ -2477,6 +2488,249 @@ Plotly.relayout = function relayout(gd, astr, val) {
+        }
+
+        if(layout) {
+            for(j = 0; j < basePlotModules.length; j++) {
+                if(basePlotModules[j].transitionAxes) {
+                    basePlotModules[j].transitionAxes(gd, layout, transitionOpts);
+                }
+            }
+        }
+
+        if(!transitionOpts.leadingEdgeRestyle) {
+            return new Promise(function(resolve, reject) {
+                completion = resolve;
+                completionTimeout = setTimeout(resolve, transitionOpts.duration);
+            });
+        }
+    }*/
+};
+
+/**
+ * Animate to a keyframe
+ *
+ * @param {string} name
+ *      name of the keyframe to create
+ * @param {object} transitionConfig
+ *      configuration for transition
+ */
+Plotly.animate = function(gd, name /*, transitionConfig*/) {
+    gd = getGraphDiv(gd);
+
+    var _frames = gd._frameData._frames;
+
+    if(!_frames[name]) {
+        Lib.warn('animateToFrame failure: keyframe does not exist', name);
+        return Promise.reject();
+    }
+
+    return Promise.resolve();
+};
+
+/**
+ * Create new keyframes
+ *
+ * @param {array of objects} frameList
+ *      list of frame definitions, in which each object includes any of:
+ *      - name: {string} name of keyframe to add
+ *      - data: {array of objects} trace data
+ *      - layout {object} layout definition
+ *      - traces {array} trace indices
+ *      - baseFrame {string} name of keyframe from which this keyframe gets defaults
+ */
+Plotly.addFrames = function(gd, frameList, indices) {
+    gd = getGraphDiv(gd);
+
+    var i, frame, j, idx;
+    var _frames = gd._frameData._frames;
+    var _hash = gd._frameData._frameHash;
+
+
+    if(!Array.isArray(frameList)) {
+        Lib.warn('addFrames failure: frameList must be an Array of frame definitions', frameList);
+        return Promise.reject();
+    }
+
+    // Create a sorted list of insertions since we run into lots of problems if these
+    // aren't in ascending order of index:
+    //
+    // Strictly for sorting. Make sure this is guaranteed to never collide with any
+    // already-exisisting indices:
+    var bigIndex = _frames.length + frameList.length * 2;
+
+    var insertions = [];
+    for(i = frameList.length - 1; i >= 0; i--) {
+        insertions.push({
+            frame: frameList[i],
+            index: (indices && indices[i] !== undefined) ? indices[i] : bigIndex + i
+        });
+    }
+
+    // Sort this, taking note that undefined insertions end up at the end:
+    insertions.sort(function(a, b) {
+        if(a.index > b.index) return -1;
+        if(a.index < b.index) return 1;
+        return 0;
+    });
+
+    var ops = [];
+    var revops = [];
+    var frameCount = _frames.length;
+
+    for(i = insertions.length - 1; i >= 0; i--) {
+        //frame = frameList[i];
+        frame = insertions[i].frame;
+
+        if(!frame.name) {
+            while(_frames[(frame.name = 'frame ' + gd._frameData._counter++)]);
```

Comment:
This lines makes my brain hurt.

I'd vote for `break`ing out of a for-loop instead.